compiler/src/iree/compiler/Codegen/Common/ExtractAddressComputationPass.cpp

```diff
@@ -0,0 +1,175 @@
+// =>
+// %new_base = subview %base[%off0,.., %offN][1,..,1][1,..,1]
+// %ld = memref.load %new_base[0,..,0] :
+//    memref<1x..x1xTy, strided<[1,..,1], offset: ?>>
+//
+// `getSrcMemRef` returns the source memref for the given load-like operation.
+//
+// Using the given rewriter, `rebuildOpFromAddressAndIndices` creates a new
+// LoadLikeOp that reads from srcMemRef[indices].
+// The returned operation will be used to replace loadOp.
+template <typename LoadLikeOp, Value (*getSrcMemRef)(LoadLikeOp),
+          LoadLikeOp (*rebuildOpFromAddressAndIndices)(
+              RewriterBase & /*rewriter*/, LoadLikeOp /*loadOp*/,
+              Value /*srcMemRef*/, ArrayRef<Value> /*indices*/)>
+struct LoadLikeOpRewriter : public OpRewritePattern<LoadLikeOp> {
+  using OpRewritePattern<LoadLikeOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(LoadLikeOp loadLikeOp,
+                                PatternRewriter &rewriter) const override {
+    Value srcMemRef = getSrcMemRef(loadLikeOp);
+    auto ldTy = srcMemRef.getType().cast<MemRefType>();
+    unsigned loadRank = ldTy.getRank();
+    // Don't waste compile time if there is nothing to rewrite.
+    if (loadRank == 0) return failure();
+
+    // If our load already has only zeros as indices there is nothing
+    // to do.
+    SmallVector<OpFoldResult> indices =
+        getAsOpFoldResult(loadLikeOp.getIndices());
+    if (std::all_of(indices.begin(), indices.end(),
+                    [](const OpFoldResult &opFold) {
+                      std::optional<int64_t> maybeConstant =
+                          getConstantIntValue(opFold);
+                      return maybeConstant && *maybeConstant == 0;
+                    })) {
+      return failure();
+    }
+
+    // Create the array of ones of the right size.
+    SmallVector<OpFoldResult> ones(loadRank, rewriter.getIndexAttr(1));
+    Location loc = loadLikeOp.getLoc();
+    auto subview = rewriter.create<memref::SubViewOp>(
+        loc, /*source=*/srcMemRef,
+        /*offsets=*/indices,
+        // TODO: Get the right sizes for non-unary loads.
+        /*sizes=*/ones, /*strides=*/ones);
+    // Rewrite the load with the subview as the base pointer.
+    SmallVector<Value> zeros(loadRank,
+                             rewriter.create<arith::ConstantIndexOp>(loc, 0));
+    LoadLikeOp newLoad = rebuildOpFromAddressAndIndices(
+        rewriter, loadLikeOp, subview.getResult(), zeros);
+    rewriter.replaceOp(loadLikeOp, newLoad.getResult());
+    return success();
+  }
+};
+
+// For now the populate method are not exposed to the rest of the world.
+static void populateExtractAddressComputationPatterns(
+    RewritePatternSet &patterns) {
+  patterns.add<
+      LoadLikeOpRewriter<memref::LoadOp,
+                         /*getSrcMemRef=*/getLoadOpSrcMemRef,
+                         /*rebuildOpFromAddressAndIndices=*/rebuildLoadOp>>(
+      patterns.getContext());
+}
+
+static void populateExtractAddressComputationGPUPatterns(
```

Comment:
Sorry for being annoying about this, but could we move these patterns into `LLVMGPU`. No other backends need this (or can use this AFAIK. I dont know SPIR-V use these as well. If that is the case, then it is fine here).

Reply:
Sure thing!
I'll follow your direction, I don't yet understand the directory layout in IREE.